File: api/internal/service/activity_service.go
package service

import (
	"strings"

	"github.com/idoceb00/elogap-api/internal/domain"
	"github.com/idoceb00/elogap-api/internal/repository"
)

type ActivityService struct {
	repo repository.ActivityRepository
}

type ListActivitiesFilter struct {
	Result   *domain.MatchResult
	Champion *string
}

func NewActivityService(repo repository.ActivityRepository) *ActivityService {
	return &ActivityService{repo: repo}
}

func (s *ActivityService) List(filter ListActivitiesFilter) ([]domain.Activity, error) {
	items, err := s.repo.List()
	if err != nil {
		return nil, err
	}

	out := make([]domain.Activity, 0, len(items))
	for _, a := range items {
		if filter.Result != nil && a.Result != *filter.Result {
			continue
		}
		if filter.Champion != nil && !strings.EqualFold(a.Champion, *filter.Champion) {
			continue
		}
		out = append(out, a)
	}
	return out, nil
}

func (s *ActivityService) FindByID(id string) (*domain.Activity, error) {
	return s.repo.FindByID(id)
}

File: api/src/main/java/com/idoceb00/elogap/service/ActivityService.java
package com.idoceb00.elogap.service;

import com.idoceb00.elogap.domain.Activity;
import com.idoceb00.elogap.domain.MatchResult;
import com.idoceb00.elogap.repository.ActivityRepository;
import org.springframework.stereotype.Service;

import java.util.List;
import java.util.Optional;
import java.util.stream.Stream;

@Service
public class ActivityService {
    private final ActivityRepository repository;

    public ActivityService(ActivityRepository repository) {
        this.repository = repository;
    }

    public List<Activity> list(Optional<MatchResult> result, Optional<String> champion) {
        Stream<Activity> stream = repository.list().stream();

        if (result.isPresent()) {
            stream = stream.filter(a -> a.result() == result.get());
        }

        if (champion.isPresent()) {
            stream = stream.filter(a -> a.champion().equalsIgnoreCase(champion.get()));
        }

        return stream.toList();
    }

    public Optional<Activity> findById(String id) {
        return repository.findById(id);
    }
}
